File: Minimal/tests/Brontide.Minimal.Host.Tests/ComponentAvailabilityBaselineTests.fs
namespace Brontide.Minimal.Host.Tests

open System
open System.IO
open System.Text.Json
open System.Threading
open System.Threading.Tasks
open Brontide.Minimal.Host
open NUnit.Framework

[<TestFixture>]
type ComponentAvailabilityBaselineTests() =
    let multiple action = Assert.Multiple(Action action)

    let start = DateTimeOffset.FromUnixTimeSeconds 1786230000L
    let interval = TimeSpan.FromSeconds 60.0

    let textValue (value: JsonElement) = value.GetString() |> Option.ofObj |> Option.defaultValue ""

    let fixture () = JsonDocument.Parse(File.ReadAllText(Path.Combine(
        TestContext.CurrentContext.TestDirectory, "component-management", "fixtures",
        "cbi65-availability-baseline-vectors.json")))

    let vectorNamed (document: JsonDocument) name =
        document.RootElement.GetProperty("vectors").EnumerateArray()
        |> Seq.find (fun vector -> textValue (vector.GetProperty "name") = name)

    let instantOf index = start + interval * float (index: int)

    /// Builds the snapshot a vector describes. The instants are the ones a cadence on the shared
    /// schedule prepares, so a vector never names a time the journal could not have recorded.
    let snapshotOf (vector: JsonElement) : ProviderTrustCadenceJournalSnapshot =
        let cycles =
            vector.GetProperty("cycles").EnumerateArray()
            |> Seq.mapi (fun index code ->
                { Index = index; Instant = instantOf index; Code = textValue code }
                : ProviderTrustCadenceJournalCycle)
            |> List.ofSeq
        { RunIdentity = ProviderTrustCadenceRunId.create "cbi65-run"
          Code = "durable-cadence-established"
          Phase = "waiting"
          MaximumCycles = 8
          Interval = interval
          PreparedInstant = instantOf cycles.Length
          Cycles = cycles
          Gaps = List.replicate (max (cycles.Length - 1) 0) interval
          NextCycleIndex = cycles.Length
          InterruptionCount = 0
          RetryCount = 0
          Cursor = None }

    let expectedOf (vector: JsonElement) =
        let cycle = vector.GetProperty "baselineCycle"
        textValue (vector.GetProperty "code"),
        (if cycle.ValueKind = JsonValueKind.Null then None else Some(instantOf (cycle.GetInt32())))

    let journalAt path =
        let opened =
            DurableProviderTrustCadenceJournal.Establish(
                path,
                ProviderTrustCadenceRunId.create "cbi65-run",
                ProviderServingTrustCadenceSchedule.create 8 interval,
                start)
        opened.Journal.Value

    let deleteTree path = try Directory.Delete(path, true) with _ -> ()

    let poll code lastAttemptCode : ProviderPublisherTrustPolicyPollResult =
        { Code = code; LastAttemptCode = lastAttemptCode; Attempts = 1; Delays = []
          AppliedSequences = []; RetainedSequences = []; Current = None
          Floor = Unchecked.defaultof<ProviderPublisherTrustPolicyRecoveryFloor> }

    /// One availability-governed cycle over a scripted policy endpoint and an empty serving set, which
    /// keeps CBI49's decision and its deadline real without launching a provider. The returned list
    /// collects every availability observation the cycle reported.
    let cadenceOver (script: string list) (baseline: DateTimeOffset option) =
        let observed = ResizeArray<ProviderTrustCycleAvailability>()
        let mutable index = 0
        let policyCycle: ProviderPublisherTrustPolicyCycle =
            fun _ _ ->
                let name = script[min index (script.Length - 1)]
                if name = "current" then poll "policy-poll-current" (Some "policy-distribution-current")
                else poll "policy-poll-exhausted" (Some "policy-distribution-transport-failed")
                |> Task.FromResult
        let sweep: ProviderServingTrustSweepCycle = fun _ -> Task.FromResult None
        let enforce =
            ProviderAvailabilityTrustCycle.enforcement
                (ProviderTrustOfflinePolicy.create (TimeSpan.FromMinutes 5.0) (TimeSpan.FromMinutes 1.0))
                (fun _ -> Task.FromResult [])
                "offline availability withdrawn"
        let inner =
            ProviderAvailabilityTrustCycle.resume baseline
                (ProviderServingTrustCycle.create policyCycle sweep) enforce
        let cycle: ProviderServingTrustCycle =
            fun now cancellationToken -> task {
                let! result = inner now cancellationToken
                result.Availability |> Option.iter observed.Add
                index <- index + 1
                return result
            }
        cycle, observed

    /// Drives the journal the way a host does — begin, run, commit, complete the gap — so the recorded
    /// instants are the ones the cycles actually ran at.
    let advance (journal: DurableProviderTrustCadenceJournal) (cycle: ProviderServingTrustCycle) count =
        task {
            for _ in 1..count do
                let instant = journal.Snapshot.PreparedInstant
                Assert.That(journal.BeginCycle(None).Code, Is.EqualTo "durable-cadence-cycle-started")
                let! result = cycle instant CancellationToken.None
                let committed = journal.CommitCycle result.Code
                if committed.Snapshot.Phase = "waiting" then
                    journal.CompleteGap(instant + interval) |> ignore
        }

    let outage baseline instant =
        task {
            let cycle, observed = cadenceOver [ "transport" ] baseline
            let! _ = cycle instant CancellationToken.None
            return Seq.tryHead observed
        }

    [<Test>]
    member _.``CBI65 C8 minimal derives the shared availability baselines``() =
        use document = fixture ()
        for vector in document.RootElement.GetProperty("vectors").EnumerateArray() do
            let actual = ProviderTrustCadenceAvailabilityRecovery.derive (snapshotOf vector)
            Assert.That(
                (actual.Code, actual.Instant), Is.EqualTo(expectedOf vector),
                textValue (vector.GetProperty "name"))

    [<Test>]
    member _.``CBI65 C1 deriving a baseline writes nothing``() =
        let root = Path.Combine(Path.GetTempPath(), $"brontide-cbi65-{Guid.NewGuid():N}")
        let path = Path.Combine(root, "cadence.bin")
        try
            let journal = journalAt path
            Assert.That(journal.BeginCycle(None).Code, Is.EqualTo "durable-cadence-cycle-started")
            Assert.That(
                (journal.CommitCycle ProviderServingTrustCycleCodes.Current).Code,
                Is.EqualTo "durable-cadence-cycle-committed")
            let before = File.ReadAllBytes path
            let derived = ProviderTrustCadenceAvailabilityRecovery.derive journal.Snapshot
            // A refused derivation must be silent too, which a synthesised snapshot is the only way to
            // reach: the journal cannot produce the record that provokes it.
            let refused =
                ProviderTrustCadenceAvailabilityRecovery.derive
                    { journal.Snapshot with
                        Cycles =
                            [ { Index = 0; Instant = start
                                Code = ProviderServingTrustCycleCodes.Stopped } ] }
            multiple (fun () ->
                Assert.That(derived.Code, Is.EqualTo "cadence-baseline-derived")
                Assert.That(refused.Code, Is.EqualTo "cadence-baseline-observation-invalid")
                Assert.That(File.ReadAllBytes path, Is.EqualTo(box before)))
        finally
            deleteTree root

    [<Test>]
    member _.``CBI65 C2 the vocabulary answers for every code it holds``() =
        use document = fixture ()
        let classification = document.RootElement.GetProperty "classification"
        multiple (fun () ->
            for code in ProviderServingTrustCycleCodes.all do
                let expected = classification.GetProperty code
                let answer = ProviderServingTrustCycleCodes.establishes code
                Assert.That(
                    answer,
                    Is.EqualTo(
                        if expected.ValueKind = JsonValueKind.Null then None
                        else Some(expected.GetBoolean())),
                    code)
                // Every code the vocabulary answers for is one a cadence may continue after, which is
                // what makes the unanswered ones unreachable in a record CBI48 wrote.
                if Option.isSome answer then
                    Assert.That(ProviderServingTrustCycleCodes.continues code, Is.True, code))

    /// The derivation reproduces the instant the live cadence held rather than a value that merely
    /// looks plausible: the same run is executed against a real journal and the two are compared.
    [<Test>]
    member _.``CBI65 C2 a replayed run yields the baseline the live cadence held``() = task {
        let root = Path.Combine(Path.GetTempPath(), $"brontide-cbi65-{Guid.NewGuid():N}")
        try
            let journal = journalAt (Path.Combine(root, "cadence.bin"))
            let cycle, observed = cadenceOver [ "current"; "current"; "transport"; "transport" ] None
            do! advance journal cycle 4
            let derived = ProviderTrustCadenceAvailabilityRecovery.derive journal.Snapshot
            multiple (fun () ->
                // The live cadence anchored its deadline on the second current cycle.
                Assert.That(derived.Instant, Is.EqualTo(Some(instantOf 1)))
                Assert.That(observed, Has.Count.EqualTo 2)
                for availability in observed do
                    Assert.That(
                        availability.Deadline,
                        Is.EqualTo(derived.Instant |> Option.map (fun value -> value.AddMinutes 5.0))))
        finally
            deleteTree root }

    [<Test>]
    member _.``CBI65 C3 the baseline does not depend on the run or its terminal code``() =
        use document = fixture ()
        let snapshot = snapshotOf (vectorNamed document "an-outage-does-not-move-it")
        let ended =
            { snapshot with
                RunIdentity = ProviderTrustCadenceRunId.create "cbi65-some-earlier-run"
                Phase = "terminal"
                Code = "durable-cadence-complete" }
        multiple (fun () ->
            // A host that shut down cleanly holds the same fact as one that crashed; refusing the
            // completed run would stop service at its first outage for no gain.
            Assert.That(
                ProviderTrustCadenceAvailabilityRecovery.derive ended,
                Is.EqualTo(ProviderTrustCadenceAvailabilityRecovery.derive snapshot))
            Assert.That(
                (ProviderTrustCadenceAvailabilityRecovery.derive ended).Code,
                Is.EqualTo "cadence-baseline-derived"))

    [<Test>]
    member _.``CBI65 C4 a record with no establishing cycle yields no instant``() =
        use document = fixture ()
        for name in [ "a-run-that-never-reached-the-endpoint-has-none"; "an-empty-record-has-none" ] do
            let actual =
                ProviderTrustCadenceAvailabilityRecovery.derive (snapshotOf (vectorNamed document name))
            multiple (fun () ->
                Assert.That(actual.Code, Is.EqualTo("cadence-baseline-absent"), name)
                Assert.That(actual.Instant, Is.EqualTo(None: DateTimeOffset option), name))

    [<Test>]
    member _.``CBI65 C5 an attempt in flight changes no derivation``() =
        let root = Path.Combine(Path.GetTempPath(), $"brontide-cbi65-{Guid.NewGuid():N}")
        try
            let journal = journalAt (Path.Combine(root, "cadence.bin"))
            journal.BeginCycle None |> ignore
            journal.CommitCycle ProviderServingTrustCycleCodes.Current |> ignore
            journal.CompleteGap(start + interval) |> ignore
            let committed = ProviderTrustCadenceAvailabilityRecovery.derive journal.Snapshot
            Assert.That(journal.BeginCycle(None).Code, Is.EqualTo "durable-cadence-cycle-started")
            let inFlight = ProviderTrustCadenceAvailabilityRecovery.derive journal.Snapshot
            multiple (fun () ->
                Assert.That(journal.Snapshot.Phase, Is.EqualTo "in-flight")
                Assert.That(inFlight, Is.EqualTo committed))
        finally
            deleteTree root

    /// C6 claims CBI48 cannot place an unclassifiable observation in front of another. That is a claim
    /// about a dependency, so it is probed: every continuing code keeps the run going and the first
    /// non-continuing one ends it in the same write.
    [<Test>]
    member _.``CBI65 C6 the journal never records an unclassifiable observation before another``() =
        for code in ProviderServingTrustCycleCodes.all do
            let root = Path.Combine(Path.GetTempPath(), $"brontide-cbi65-{Guid.NewGuid():N}")
            try
                let journal = journalAt (Path.Combine(root, "cadence.bin"))
                journal.BeginCycle None |> ignore
                let committed = journal.CommitCycle code
                let unanswered = Option.isNone (ProviderServingTrustCycleCodes.establishes code)
                let expected =
                    if unanswered then "cadence-baseline-observation-invalid"
                    elif ProviderServingTrustCycleCodes.establishes code = Some true then
                        "cadence-baseline-derived"
                    else "cadence-baseline-absent"
                multiple (fun () ->
                    Assert.That(
                        committed.Snapshot.Phase,
                        Is.EqualTo(if unanswered then "terminal" else "waiting"), code)
                    // A terminal journal accepts nothing further, so no later observation can follow
                    // the one the derivation could not classify.
                    if unanswered then
                        Assert.That(journal.BeginCycle(None).Code, Does.StartWith "durable-cadence-", code)
                    Assert.That(
                        (ProviderTrustCadenceAvailabilityRecovery.derive journal.Snapshot).Code,
                        Is.EqualTo(expected), code))
            finally
                deleteTree root

    /// The composed effect. Three successors run the same outage cycle at the same instant and differ
    /// only in the baseline they start from: the derived one, none at all, and the restart instant —
    /// the tempting wrong answer, which renews grace on every restart so a crash loop never expires.
    [<Test>]
    member _.``CBI65 C7 a resumed cadence continues the outage it was in``() = task {
        let root = Path.Combine(Path.GetTempPath(), $"brontide-cbi65-{Guid.NewGuid():N}")
        try
            let journal = journalAt (Path.Combine(root, "cadence.bin"))
            let cycle, before = cadenceOver [ "current"; "transport"; "transport" ] None
            do! advance journal cycle 3
            let derived = ProviderTrustCadenceAvailabilityRecovery.derive journal.Snapshot
            let restart = journal.Snapshot.PreparedInstant
            let! resumed = outage derived.Instant restart
            let! none = outage None restart
            let! renewed = outage (Some restart) restart
            let interrupted = (Seq.last before).Deadline
            multiple (fun () ->
                // The outage the host was in is the outage it comes back to.
                Assert.That(resumed |> Option.bind _.Deadline, Is.EqualTo interrupted)
                Assert.That(resumed |> Option.bind _.DecisionCode, Is.EqualTo(Some "offline-idle"))
                // Without a baseline the run stops service instead, which is CBI64's stated limit and
                // what this slice removes.
                Assert.That(
                    none |> Option.bind _.DecisionCode,
                    Is.EqualTo(Some "offline-service-stop-required"))
                // Anchoring on the restart moves the deadline forward, so an outage spanning restarts
                // would never expire.
                Assert.That(renewed |> Option.bind _.Deadline |> Option.get, Is.GreaterThan(Option.get interrupted)))
        finally
            deleteTree root }

File: Reference/tests/Brontide.Reference.Studio.Tests/ComponentAvailabilityBaselineTests.cs
using System.Text.Json;

namespace Brontide.Reference.Studio.Tests;

public sealed partial class ComponentBindingIntegrationTests
{
    private static readonly DateTimeOffset Cbi65Start = DateTimeOffset.FromUnixTimeSeconds(1_786_230_000);
    private static readonly TimeSpan Cbi65Interval = TimeSpan.FromSeconds(60);

    private static JsonDocument Cbi65Fixture() => JsonDocument.Parse(File.ReadAllText(Path.Combine(
        TestContext.CurrentContext.TestDirectory, "component-management", "fixtures",
        "cbi65-availability-baseline-vectors.json")));

    private static JsonElement Cbi65Vector(JsonDocument fixture, string name) =>
        fixture.RootElement.GetProperty("vectors").EnumerateArray()
            .Single(vector => vector.GetProperty("name").GetString() == name);

    /// <summary>
    /// Builds the snapshot a vector describes. The instants are the ones a cadence on the shared
    /// schedule prepares, so a vector never names a time the journal could not have recorded.
    /// </summary>
    private static ProviderTrustCadenceJournalSnapshot Cbi65Snapshot(JsonElement vector)
    {
        var cycles = vector.GetProperty("cycles").EnumerateArray()
            .Select((code, index) => new ProviderTrustCadenceJournalCycle(
                index, Cbi65Start + Cbi65Interval * index, code.GetString()!))
            .ToArray();
        return new(
            ProviderTrustCadenceRunId.Create("cbi65-run"), "durable-cadence-established", "waiting",
            8, Cbi65Interval, Cbi65Start + Cbi65Interval * cycles.Length, cycles,
            Enumerable.Repeat(Cbi65Interval, Math.Max(cycles.Length - 1, 0)).ToArray(),
            cycles.Length, 0, 0);
    }

    private static (string Code, DateTimeOffset? Instant) Cbi65Expected(JsonElement vector)
    {
        var cycle = vector.GetProperty("baselineCycle");
        return (vector.GetProperty("code").GetString()!,
            cycle.ValueKind == JsonValueKind.Null
                ? null
                : Cbi65Start + Cbi65Interval * cycle.GetInt32());
    }

    [Test]
    public void Shared_cbi65_vectors_derive_the_availability_baseline()
    {
        using var fixture = Cbi65Fixture();
        foreach (var vector in fixture.RootElement.GetProperty("vectors").EnumerateArray())
        {
            var actual = ProviderTrustCadenceAvailabilityRecovery.Derive(Cbi65Snapshot(vector));
            Assert.That((actual.Code, actual.Instant), Is.EqualTo(Cbi65Expected(vector)),
                $"vector {vector.GetProperty("name").GetString()}");
        }
    }

    [Test]
    public void Cbi65_C1_deriving_a_baseline_writes_nothing()
    {
        var root = Path.Combine(Path.GetTempPath(), $"brontide-cbi65-{Guid.NewGuid():N}");
        var path = Path.Combine(root, "cadence.bin");
        try
        {
            var journal = Cbi65Journal(path);
            Assert.That(journal.BeginCycle().Code, Is.EqualTo("durable-cadence-cycle-started"));
            Assert.That(journal.CommitCycle(ProviderServingTrustCycleCodes.Current).Code,
                Is.EqualTo("durable-cadence-cycle-committed"));

            var before = File.ReadAllBytes(path);
            var derived = ProviderTrustCadenceAvailabilityRecovery.Derive(journal.Snapshot);
            // A refused derivation must be silent too, which a synthesised snapshot is the only way
            // to reach: the journal cannot produce the record that provokes it.
            var refused = ProviderTrustCadenceAvailabilityRecovery.Derive(
                journal.Snapshot with
                {
                    Cycles = [new(0, Cbi65Start, ProviderServingTrustCycleCodes.Stopped)],
                });
            Assert.Multiple(() =>
            {
                Assert.That(derived.Code, Is.EqualTo("cadence-baseline-derived"));
                Assert.That(refused.Code, Is.EqualTo("cadence-baseline-observation-invalid"));
                Assert.That(File.ReadAllBytes(path), Is.EqualTo(before));
            });
        }
        finally
        {
            Cbi32DeleteTree(root);
        }
    }

    [Test]
    public void Cbi65_C2_the_vocabulary_answers_for_every_code_it_holds()
    {
        using var fixture = Cbi65Fixture();
        var classification = fixture.RootElement.GetProperty("classification");
        Assert.Multiple(() =>
        {
            foreach (var code in ProviderServingTrustCycleCodes.All)
            {
                var expected = classification.GetProperty(code);
                Assert.That(ProviderServingTrustCycleCodes.Establishes(code),
                    Is.EqualTo(expected.ValueKind == JsonValueKind.Null ? null : expected.GetBoolean()),
                    code);
                // Every code the vocabulary answers for is one a cadence may continue after, which is
                // what makes the unanswered ones unreachable in a record CBI48 wrote.
                if (ProviderServingTrustCycleCodes.Establishes(code) is not null)
                    Assert.That(ProviderServingTrustCycleCodes.Continues(code), Is.True, code);
            }
        });
    }

    /// <summary>
    /// The derivation reproduces the instant the live cadence held rather than a value that merely
    /// looks plausible: the same run is executed against a real journal and the two are compared.
    /// </summary>
    [Test]
    public async Task Cbi65_C2_a_replayed_run_yields_the_baseline_the_live_cadence_held()
    {
        var root = Path.Combine(Path.GetTempPath(), $"brontide-cbi65-{Guid.NewGuid():N}");
        try
        {
            var journal = Cbi65Journal(Path.Combine(root, "cadence.bin"));
            var (cycle, observed) = Cbi65Cadence(
                ["current", "current", "transport", "transport"], baseline: null);
            await Cbi65AdvanceAsync(journal, cycle, 4);

            var derived = ProviderTrustCadenceAvailabilityRecovery.Derive(journal.Snapshot);
            Assert.Multiple(() =>
            {
                // The live cadence anchored its deadline on the second current cycle.
                Assert.That(derived.Instant, Is.EqualTo(Cbi65Start + Cbi65Interval));
                Assert.That(observed, Has.Count.EqualTo(2));
                foreach (var availability in observed)
                    Assert.That(availability.Deadline,
                        Is.EqualTo(derived.Instant + TimeSpan.FromMinutes(5)));
            });
        }
        finally
        {
            Cbi32DeleteTree(root);
        }
    }

    [Test]
    public void Cbi65_C3_the_baseline_does_not_depend_on_the_run_or_its_terminal_code()
    {
        using var fixture = Cbi65Fixture();
        var vector = Cbi65Vector(fixture, "an-outage-does-not-move-it");
        var snapshot = Cbi65Snapshot(vector);
        var ended = snapshot with
        {
            RunIdentity = ProviderTrustCadenceRunId.Create("cbi65-some-earlier-run"),
            Phase = "terminal",
            Code = "durable-cadence-complete",
        };
        Assert.Multiple(() =>
        {
            // A host that shut down cleanly holds the same fact as one that crashed; refusing the
            // completed run would stop service at its first outage for no gain.
            Assert.That(ProviderTrustCadenceAvailabilityRecovery.Derive(ended),
                Is.EqualTo(ProviderTrustCadenceAvailabilityRecovery.Derive(snapshot)));
            Assert.That(ProviderTrustCadenceAvailabilityRecovery.Derive(ended).Code,
                Is.EqualTo("cadence-baseline-derived"));
        });
    }

    [Test]
    public void Cbi65_C4_a_record_with_no_establishing_cycle_yields_no_instant()
    {
        using var fixture = Cbi65Fixture();
        foreach (var name in new[]
                 {
                     "a-run-that-never-reached-the-endpoint-has-none",
                     "an-empty-record-has-none",
                 })
        {
            var actual = ProviderTrustCadenceAvailabilityRecovery.Derive(
                Cbi65Snapshot(Cbi65Vector(fixture, name)));
            Assert.Multiple(() =>
            {
                Assert.That(actual.Code, Is.EqualTo("cadence-baseline-absent"), name);
                Assert.That(actual.Instant, Is.Null, name);
            });
        }
    }

    [Test]
    public void Cbi65_C5_an_attempt_in_flight_changes_no_derivation()
    {
        var root = Path.Combine(Path.GetTempPath(), $"brontide-cbi65-{Guid.NewGuid():N}");
        try
        {
            var journal = Cbi65Journal(Path.Combine(root, "cadence.bin"));
            journal.BeginCycle();
            journal.CommitCycle(ProviderServingTrustCycleCodes.Current);
            journal.CompleteGap(Cbi65Start + Cbi65Interval);
            var committed = ProviderTrustCadenceAvailabilityRecovery.Derive(journal.Snapshot);

            Assert.That(journal.BeginCycle().Code, Is.EqualTo("durable-cadence-cycle-started"));
            var inFlight = ProviderTrustCadenceAvailabilityRecovery.Derive(journal.Snapshot);
            Assert.Multiple(() =>
            {
                Assert.That(journal.Snapshot.Phase, Is.EqualTo("in-flight"));
                Assert.That(inFlight, Is.EqualTo(committed));
            });
        }
        finally
        {
            Cbi32DeleteTree(root);
        }
    }

    /// <summary>
    /// C6 claims CBI48 cannot place an unclassifiable observation in front of another. That is a claim
    /// about a dependency, so it is probed: every continuing code keeps the run going and the first
    /// non-continuing one ends it in the same write.
    /// </summary>
    [Test]
    public void Cbi65_C6_the_journal_never_records_an_unclassifiable_observation_before_another()
    {
        foreach (var code in ProviderServingTrustCycleCodes.All)
        {
            var root = Path.Combine(Path.GetTempPath(), $"brontide-cbi65-{Guid.NewGuid():N}");
            try
            {
                var journal = Cbi65Journal(Path.Combine(root, "cadence.bin"));
                journal.BeginCycle();
                var committed = journal.CommitCycle(code);
                var unanswered = ProviderServingTrustCycleCodes.Establishes(code) is null;
                Assert.Multiple(() =>
                {
                    Assert.That(committed.Snapshot.Phase,
                        Is.EqualTo(unanswered ? "terminal" : "waiting"), code);
                    // A terminal journal accepts nothing further, so no later observation can follow
                    // the one the derivation could not classify.
                    if (unanswered)
                        Assert.That(journal.BeginCycle().Code, Does.StartWith("durable-cadence-"), code);
                    Assert.That(ProviderTrustCadenceAvailabilityRecovery.Derive(journal.Snapshot).Code,
                        Is.EqualTo(unanswered
                            ? "cadence-baseline-observation-invalid"
                            : ProviderServingTrustCycleCodes.Establishes(code) == true
                                ? "cadence-baseline-derived"
                                : "cadence-baseline-absent"), code);
                });
            }
            finally
            {
                Cbi32DeleteTree(root);
            }
        }
    }

    /// <summary>
    /// The composed effect. Three successors run the same outage cycle at the same instant and differ
    /// only in the baseline they start from: the derived one, none at all, and the restart instant —
    /// the tempting wrong answer, which renews grace on every restart so a crash loop never expires.
    /// </summary>
    [Test]
    public async Task Cbi65_C7_a_resumed_cadence_continues_the_outage_it_was_in()
    {
        var root = Path.Combine(Path.GetTempPath(), $"brontide-cbi65-{Guid.NewGuid():N}");
        try
        {
            var journal = Cbi65Journal(Path.Combine(root, "cadence.bin"));
            var (first, before) = Cbi65Cadence(["current", "transport", "transport"], baseline: null);
            await Cbi65AdvanceAsync(journal, first, 3);
            var derived = ProviderTrustCadenceAvailabilityRecovery.Derive(journal.Snapshot);

            var restart = journal.Snapshot.PreparedInstant;
            var resumed = await Cbi65OutageAsync(derived.Instant, restart);
            var none = await Cbi65OutageAsync(null, restart);
            var renewed = await Cbi65OutageAsync(restart, restart);

            Assert.Multiple(() =>
            {
                // The outage the host was in is the outage it comes back to.
                Assert.That(resumed!.Deadline, Is.EqualTo(before[^1].Deadline));
                Assert.That(resumed.DecisionCode, Is.EqualTo("offline-idle"));
                // Without a baseline the run stops service instead, which is CBI64's stated limit and
                // what this slice removes.
                Assert.That(none!.DecisionCode, Is.EqualTo("offline-service-stop-required"));
                // Anchoring on the restart moves the deadline forward, so an outage spanning restarts
                // would never expire.
                Assert.That(renewed!.Deadline, Is.GreaterThan(before[^1].Deadline!.Value));
            });
        }
        finally
        {
            Cbi32DeleteTree(root);
        }
    }

    private static async Task<ProviderTrustCycleAvailability?> Cbi65OutageAsync(
        DateTimeOffset? baseline, DateTimeOffset instant)
    {
        var (cycle, observed) = Cbi65Cadence(["transport"], baseline);
        await cycle.RunAsync(instant, CancellationToken.None);
        return observed.SingleOrDefault();
    }

    private static DurableProviderTrustCadenceJournal Cbi65Journal(string path) =>
        DurableProviderTrustCadenceJournal.Establish(
            path, ProviderTrustCadenceRunId.Create("cbi65-run"),
            ProviderServingTrustCadenceSchedule.Create(8, Cbi65Interval), Cbi65Start).Journal!;

    /// <summary>
    /// One availability-governed cycle over a scripted policy endpoint and an empty serving set, which
    /// keeps CBI49's decision and its deadline real without launching a provider. The returned list
    /// collects every availability observation the cycle reported.
    /// </summary>
    private static (IProviderServingTrustCycle Cycle, List<ProviderTrustCycleAvailability> Observed)
        Cbi65Cadence(IReadOnlyList<string> script, DateTimeOffset? baseline)
    {
        var polls = script.Select(name => name == "current"
            ? ("policy-poll-current", (string?)"policy-distribution-current")
            : ("policy-poll-exhausted", "policy-distribution-transport-failed")).ToArray();
        var policy = new Cbi64PolicyCycle(polls, null!);
        var observed = new List<ProviderTrustCycleAvailability>();
        var cycle = new ProviderAvailabilityTrustCycle(
            new ProviderServingTrustCycle(policy, new Cbi64Sweep()),
            new ProviderOfflineEnforcementCycle(
                ProviderTrustOfflinePolicy.Create(TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(1)),
                _ => ValueTask.FromResult<IReadOnlyList<ProviderServingActivation>>([]),
                "offline availability withdrawn"),
            baseline);
        return (new Cbi65RecordingCycle(cycle, policy, observed), observed);
    }

    private sealed class Cbi65RecordingCycle(
        IProviderServingTrustCycle inner,
        Cbi64PolicyCycle policy,
        List<ProviderTrustCycleAvailability> observed) : IProviderServingTrustCycle
    {
        public async Task<ProviderServingTrustCycleResult> RunAsync(
            DateTimeOffset now, CancellationToken cancellationToken)
        {
            var result = await inner.RunAsync(now, cancellationToken).ConfigureAwait(false);
            if (result.Availability is { } availability) observed.Add(availability);
            policy.Cycle++;
            return result;
        }
    }

    /// <summary>
    /// Drives the journal the way a host does — begin, run, commit, complete the gap — so the recorded
    /// instants are the ones the cycles actually ran at.
    /// </summary>
    private static async Task Cbi65AdvanceAsync(
        DurableProviderTrustCadenceJournal journal,
        IProviderServingTrustCycle cycle,
        int cycles)
    {
        for (var index = 0; index < cycles; index++)
        {
            var instant = journal.Snapshot.PreparedInstant;
            Assert.That(journal.BeginCycle().Code, Is.EqualTo("durable-cadence-cycle-started"));
            var result = await cycle.RunAsync(instant, CancellationToken.None);
            var committed = journal.CommitCycle(result.Code);
            if (committed.Snapshot.Phase == "waiting")
                journal.CompleteGap(instant + Cbi65Interval);
        }
    }
}
